### DISCUSSION TITLE
avif support, required dependencies

### DISCUSSION THREAD

**Discussion by 0323pin** [woelper/oculante#136] [General]
I've gone through the commits between 0.6.41 and 0.6.50 and I'm confused as to which external dependencies are actually required on the build system.

From reading the commit log, looks like only `perl` is needed after all the _on_/_off_ rounds :)

Is this correct? Or are `dav1d`, `meson`, etc (which ones) still required.

It's still not enable the `turbo` feature by default on NetBSD. Actually, I can't update the package as I'm still waiting for Rust-1.66.1 to be merged. But, I can build it locally :)

Thanks for the clarification.

**Comment by 0323pin** [comment]
@woelper 

Quoting the ReadMe,
> If you disable turbo (on by default), the turbojpeg library will not be used to open jpeg images. You won't need Nasm to be installed.

I'm building with the following flags:
```
CARGO_NO_DEFAULT_FEATURES=	YES
CARGO_FEATURES+=		file_open
```

But, I'm still getting the following error:

```
error: failed to run custom build command for `libaom-sys v0.13.0`

Caused by:
  process didn't exit successfully: `/usr/pkgsrc/wip/oculante/work/oculante-0.6.50/target/release/build/libaom-sys-fc9ebeea2e6835f9/build-script-build` (exit status: 101)
  --- stdout
  CMAKE_TOOLCHAIN_FILE_x86_64-unknown-netbsd = None
  CMAKE_TOOLCHAIN_FILE_x86_64_unknown_netbsd = None
  HOST_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_x86_64-unknown-netbsd = None
  CMAKE_GENERATOR_x86_64_unknown_netbsd = None
  HOST_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_x86_64-unknown-netbsd = None
  CMAKE_PREFIX_PATH_x86_64_unknown_netbsd = None
  HOST_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_x86_64-unknown-netbsd = None
  CMAKE_x86_64_unknown_netbsd = None
  HOST_CMAKE = None
  CMAKE = None
  running: "cmake" "/usr/pkgsrc/wip/oculante/work/vendor/libaom-sys-0.13.0/vendor" "-DENABLE_DOCS=0" "-DENABLE_EXAMPLES=0" "-DENABLE_TESTDATA=0" "-DENABLE_TESTS=0" "-DENABLE_TOOLS=0" "-DCMAKE_INSTALL_LIBDIR=lib" "-DCMAKE_INSTALL_PREFIX=/usr/pkgsrc/wip/oculante/work/oculante-0.6.50/target/release/build/libaom-sys-c36e094bc93e2ab1/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64 -pipe -fPIC -D_FORTIFY_SOURCE=2 -I/usr/pkg/include -I/usr/include -I/usr/X11R7/include -I/usr/X11R7/include/freetype2 -I/usr/pkg/include/glib-2.0 -I/usr/pkg/include/gio-unix-2.0 -I/usr/pkg/lib/glib-2.0/include -I/usr/pkg/include/harfbuzz -I/usr/X11R7/include/libdrm" "-DCMAKE_C_COMPILER=/usr/pkgsrc/wip/oculante/work/.cwrapper/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64 -pipe -fPIC -D_FORTIFY_SOURCE=2 -I/usr/pkg/include -I/usr/include -I/usr/X11R7/include -I/usr/X11R7/include/freetype2 -I/usr/pkg/include/glib-2.0 -I/usr/pkg/include/gio-unix-2.0 -I/usr/pkg/lib/glib-2.0/include -I/usr/pkg/include/harfbuzz -I/usr/X11R7/include/libdrm" "-DCMAKE_CXX_COMPILER=/usr/pkgsrc/wip/oculante/work/.cwrapper/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64 -pipe -fPIC -D_FORTIFY_SOURCE=2 -I/usr/pkg/include -I/usr/include -I/usr/X11R7/include -I/usr/X11R7/include/freetype2 -I/usr/pkg/include/glib-2.0 -I/usr/pkg/include/gio-unix-2.0 -I/usr/pkg/lib/glib-2.0/include -I/usr/pkg/include/harfbuzz -I/usr/X11R7/include/libdrm" "-DCMAKE_ASM_COMPILER=/usr/pkgsrc/wip/oculante/work/.cwrapper/bin/cc" "-DCMAKE_BUILD_TYPE=Release"
  -- The C compiler identification is GNU 10.4.0
  -- The CXX compiler identification is GNU 10.4.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/pkgsrc/wip/oculante/work/.cwrapper/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/pkgsrc/wip/oculante/work/.cwrapper/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Found Git: /usr/pkg/bin/git (found version "2.39.0")
  -- Found Perl: /usr/pkgsrc/wip/oculante/work/.tools/bin/perl (found version "5.36.0")
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
  -- Looking for pthread_create in pthreads
  -- Looking for pthread_create in pthreads - not found
  -- Looking for pthread_create in pthread
  -- Looking for pthread_create in pthread - found
  -- Found Threads: TRUE
  -- Performing Test C_TEST_COMPILED
  -- Performing Test C_TEST_COMPILED - Failed
  -- Configuring incomplete, errors occurred!
  See also "/usr/pkgsrc/wip/oculante/work/oculante-0.6.50/target/release/build/libaom-sys-c36e094bc93e2ab1/out/build/CMakeFiles/CMakeOutput.log".
  See also "/usr/pkgsrc/wip/oculante/work/oculante-0.6.50/target/release/build/libaom-sys-c36e094bc93e2ab1/out/build/CMakeFiles/CMakeError.log".

  --- stderr
  --- aom_configure: Detected CPU: x86_64
  Running C compiler test: pie_check
  C Compiler test pie_check failed.
  CMake Error at build/cmake/aom_configure.cmake:164 (message):
    Unable to find assembler.  Install 'yasm' or 'nasm.' To build without
    optimizations, add -DAOM_TARGET_CPU=generic to your cmake command line.
  Call Stack (most recent call first):
    CMakeLists.txt:66 (include)


  thread 'main' panicked at '
  command did not execute successfully, got: exit status: 1

  build script failed, must exit now', /usr/pkgsrc/wip/oculante/work/vendor/cmake-0.1.49/src/lib.rs:1104:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
*** Error code 101

Stop.
```

**Comment by 0323pin** [comment]
Adding `nasm` to the cook makes it build successfully.

```
warning: unused import: `bail`
  --> src/utils.rs:25:22
   |
25 | use anyhow::{anyhow, bail, Context, Result};
   |                      ^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: `oculante` (bin "oculante") generated 1 warning
    Finished release [optimized] target(s) in 21m 09s
```

![2023-01-16-135446_1366x768_scrot](https://user-images.githubusercontent.com/90570748/212683398-3896ca52-757b-4958-8a49-0cb63e83b675.png)

**Comment by woelper** [comment]
Hi, and again thank you for your hard work!

I was in fact so pissed about avif that I switched to a different implementation. I even failed to build on windows. So now there should be minimal or no dependencies, so avif should not be behind a feature flag. (Basically I will try to steer around non-native rust libs, and if impossible will create individual features).

Sorry again for the confusion, and great you got it working!

**Reply by eugenesvk** [reply]
do you have a test build with dav1d I could check? The current version with `avif_decode` does't open some files that the older https://github.com/dreampiggy/AVIFQuickLook supports, so wanted to check if that's related to the encoder
(also, I've read that dav1d was much better for performance, but that was a while ago, so not sure is relevant)

**Reply by woelper** [reply]
I don't have a build for that, but it's fairly easy to try yourself:

You can clone the repo and use this commit: https://github.com/woelper/oculante/commit/2d01aec77a2fa39db4dca6efb62cf5f548b2c2f9
(`git checkout 2d01aec77a2fa39db4dca6efb62cf5f548b2c2f9`)
after this you can run `cargo build --release` to build. If you don't have rust, you can get that from rustup.rs.

I've tested with Dav1d back then, and it _did_ open more files. But as said, getting it to build on CI (github actions) was a nightmare on windows. I spent almost a whole weekend on that, and this is not worth the time. I _could_ use dav1d just on platforms that build it and fall back to another lib, but then I end up making lots of different feature flags.

**Reply by woelper** [reply]
I have revisited this and re-enabled `dav1d` behind a feature. Now I could build this with better support on linux and mac, while windows has slightly worse avif support until someone makes it run on github actions.

**Reply by woelper** [reply]
For reference, all you need to do is run `cargo build --release --features dav1d` on master now. The sad story about avif is that it adds around 5MB to the executable size...

**Reply by woelper** [reply]
Oh, I forgot that I build on ubuntu 20 on github CI, and they are missing dav1d libraries so I can't build it there too. So there would just be mac supporting it by default I guess. Once I would build this for ubuntu 22 people with older linux could not use oculante any more I believe due to glibc differences...

**Reply by 0323pin** [reply]
> Oh, I forgot that I build on ubuntu 20 on github CI, and they are missing dav1d libraries so I can't build it there too. So there would just be mac supporting it by default I guess. Once I would build this for ubuntu 22 people with older linux could not use oculante any more I believe due to glibc differences...

Couldn't you just build it static linked against `musl` and avoid `glibc` version issues?

**Reply by woelper** [reply]
That could work (currently trying `cargo build --target x86_64-unknown-linux-musl`) but I've read this might not be good for performance, so I might have to benchmark this.

**Reply by woelper** [reply]
and now every build script wants `musl-g++`... which I don't have...

**Reply by 0323pin** [reply]
Duh! Sorry, of course you need `musl` libc.
I don't know why it would be bad for performance, though. Here are some stats, http://www.etalabs.net/compare_libcs.html

I've been running my Linux system on `musl` since 2017.

**Comment by 0323pin** [comment]
> Sorry again for the confusion, and great you got it working!

🤣 you don't need a NetBSD CI, you already have one.

> avif should not be behind a feature flag.

Bring it on and we will see. Today, I had to add `perl` and `nasm` to the _cooking_. If you are curious you can have a look at the `Makefile` and `cargo-depends.mk` files which are waiting in my repository until `rust`-1.66.1 is merged, https://codeberg.org/pin/netbsd/src/branch/master/Packages/oculante-0.6.50

**Reply by woelper** [reply]
you can now happily mix and match with `avif_native` and `dav1d`. If you enable both, `dav1d` takes higher priority.

**Reply by 0323pin** [reply]
I hope we get `rust`-1.66.1 this week, I'm already holding 14 updates on my repository 😢

But, I can probably find some time to test build with both to see.

**Comment by Stoppedpuma** [comment]
This was implemented at some point so I'll close this. Thanks!
